src/Microsoft.AspNet.Mvc.ModelBinding/Validation/DefaultBodyModelValidator.cs

```diff
@@ -138,66 +130,69 @@ private bool ValidateElements(IEnumerable model, ValidationContext validationCon
                 if (element != null || anyValidatorsDefined)
                 {
                     elementMetadata.Model = element;
-
-                    if (!ValidateNonVisitedNodeAndChildren(elementMetadata, validationContext, validators))
+                    var elementKey = ModelBindingHelper.CreateIndexModelName(currentKey, index);
+                    if (!ValidateNonVisitedNodeAndChildren(elementKey, elementMetadata, validationContext, validators))
                     {
                         isValid = false;
                     }
                 }
 
-                elementScope.Index++;
+                index++;
             }
 
-            validationContext.KeyBuilders.Pop();
             return isValid;
         }
 
         // Validates a single node (not including children)
         // Returns true if validation passes successfully
         private static bool ShallowValidate(
+            string modelKey,
             ModelMetadata metadata,
             ValidationContext validationContext,
             [NotNull] IEnumerable<IModelValidator> validators)
         {
             var isValid = true;
-            string modelKey = null;
 
             // When the are no validators we bail quickly. This saves a GetEnumerator allocation.
             // In a large array (tens of thousands or more) scenario it's very significant.
             var validatorsAsCollection = validators as ICollection;
-            if (validatorsAsCollection != null && validatorsAsCollection.Count == 0)
+            if (validatorsAsCollection == null || validatorsAsCollection.Count > 0)
             {
-                return isValid;
-            }
+                var modelValidationContext =
+                        new ModelValidationContext(validationContext.ModelValidationContext, metadata);
+                var modelState = validationContext.ModelValidationContext.ModelState;
+                var fieldValidationState = modelState.GetFieldValidationState(modelKey);
```

Comment:
`fieldValidationState` it's hard for me to pin down what this means exactly. Maybe we can come up with better names?